### PR TITLE
Chaos Engineering

### PR DESCRIPTION
**Pull request by nvprotas**
Implementation of chaos-engineering:
- killing pods of components 
- injecting network faults in Istio sidecars
- running IO chaos in sidecars

### CODE REVIEW

File: src/main/java/io/microconfig/osdf/api/MainApiCaller.java [hunk 1]
```diff
@@ -41,6 +42,7 @@ public void call(List<String> args) {
                 .addImpl(ManagementApi.class, managementApi(paths, oc))
                 .addImpl(InfoApi.class, infoApi(paths, oc))
                 .addImpl(SystemApi.class, systemApi(paths))
+                .addImpl(ChaosApi.class, chaosApi(paths,oc))
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/api/MainApiCaller.java]: fixed

File: src/main/java/io/microconfig/osdf/chaos/ChaosStep.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                        getInt(chaosSetMap, "httpAbort"),
+                        getInt(chaosSetMap, "httpDelay"),
+                        getInt(chaosSetMap, "ioChaosTimeout"),
+                        getInt(chaosSetMap, "podChaosTimeout")
+                )
+        );
+    }
+
+    public void runStep(OSDFPaths paths, OCExecutor ocExecutor) {
+        long start = System.nanoTime();
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/ChaosStep.java]: fixed

File: src/main/java/io/microconfig/osdf/chaos/ChaosStep.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        if (number == null) {
+            throw new RuntimeException("Assign numbers for all steps in test plan!");
+        }
+
+        if (severity == null) {
+            throw new RuntimeException("Assign severity for step " + number);
+        }
+
+        if (severity < 0 || severity > 100) {
+            throw new RuntimeException("Incorrect severity " + severity + " in step " + number);
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/ChaosStep.java]: fixed

File: src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import io.microconfig.osdf.service.files.ServiceFiles;
+import io.microconfig.osdf.service.istio.IstioService;
+import lombok.AllArgsConstructor;
+
+import static io.microconfig.osdf.utils.IstioUtils.toIstioService;
+import static io.microconfig.utils.Logger.announce;
+
+
+@AllArgsConstructor
+public class NetworkChaosDeployer implements ServiceDeployer {
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java:15]: why is this a ServiceDeployer?

File: configs/repo/components/osdf/osdf-settings/openshift-urls/deploy.deploy [hunk 5]
```diff
@@ -1,4 +1,5 @@
-cluster.url.base: unknown
-cluster.url.api: unknown
-project: unknown
-registry.project.url: unknown
+cluster.url.base: unknown1
+cluster.url.api: https://api.us-west-1.starter.openshift-online.com:6443
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply configs/repo/components/osdf/osdf-settings/openshift-urls/deploy.deploy]: fixed

File: src/main/java/io/microconfig/osdf/chaos/ChaosSet.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import lombok.Getter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.microconfig.osdf.istio.faults.Fault.fault;
+
+@Getter
+@AllArgsConstructor
+public class ChaosSet {
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/ChaosSet.java]: fixed

File: src/main/java/io/microconfig/osdf/chaos/ChaosStep.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+import static io.microconfig.osdf.chaos.ChaosSet.chaosSet;
+import static io.microconfig.osdf.utils.ThreadUtils.sleepSec;
+import static io.microconfig.osdf.utils.YamlUtils.*;
+import static io.microconfig.utils.Logger.announce;
+import static java.lang.System.nanoTime;
+
+@AllArgsConstructor
+public class ChaosStep {
+    private final Integer number;
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/ChaosStep.java]: fixed

File: src/main/java/io/microconfig/osdf/chaos/chaosRunners/IOChaosRunner.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        return new IOChaosRunner(paths, oc);
+    }
+
+    @Override
+    public void run(List<String> components, ChaosSet chaosSet, Integer severity, Integer duration) {
+        List<ServiceDeployPack> deployPacks = defaultServiceDeployPacksLoader(paths, components, cli).loadPacks();
+        deployPacks.forEach(pack -> startStress2(pack, severity, chaosSet.getIoStressTimeout()));
+    }
+
+    private void startStress2(ServiceDeployPack pack, int chaosChance, int timeout) {
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/chaosRunners/IOChaosRunner.java]: fixed

File: src/main/java/io/microconfig/osdf/cluster/pod/Pod.java [hunk 9]
```diff
@@ -50,6 +50,11 @@ public void delete() {
                 .throwExceptionIfError();
     }
 
+    public void forceDelete() {
+        cli.execute("oc delete pod " + name + " --grace-period=0 --force")
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/cluster/pod/Pod.java]: fixed

File: src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java (hunk 4 above)
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java]: fixed

File: src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java]: fixed

File: src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    private final Set<Chaos> chaosSet;
+
+    public static ChaosExperiment chaosExperiment(OSDFPaths paths, ClusterCLI cli, ChaosComponent component) {
+        Map<String, Object> rules = getMap(loadFromPath(component.getPathToPlan()), "rules");
+        Set<Chaos> chaosSet = rules.entrySet().stream().map(Chaos::chaos).collect(Collectors.toSet());
+        check(chaosSet);
+        return new ChaosExperiment(paths, cli, chaosSet);
+    }
+
+    public static void check(Set<Chaos> chaosSet) {
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java:35]: do not use static methods for logic

File: src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                thread.join();
+            } catch (InterruptedException e) {
+                currentThread().interrupt();
+                e.printStackTrace();
+            }
+        });
+        error("All chaos threads joined");
+    }
+
+    Runnable toRunnable(Chaos chaos) {
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java:78]: should be private?

File: src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        podAndIOChaosIntersectionCheck(chaosSet);
+    }
+
+    public static void stop(OSDFPaths paths, ClusterCLI cli) {
+        Chaos.getAllChaosImpls().forEach(chaos -> chaos.stop(paths, cli));
+    }
+
+    public void run() {
+        announce("Launch of chaos");
+        Set<Runnable> runnables = chaosSet.stream().map(this::toRunnable).collect(Collectors.toUnmodifiableSet());
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java:48]: Why does every chaos run in thread?

File: src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        Set<Thread> threads = runnables.stream().map(Thread::new).collect(Collectors.toUnmodifiableSet());
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/chaos/ChaosExperiment.java:49]: Use static imports

File: src/main/java/io/microconfig/osdf/chaos/types/Chaos.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+            case "network":
+                return networkChaos(name, yaml);
+            case "pod":
+                return podChaos(name, yaml);
+            default:
+                throw new OSDFException("Unknown type of chaos: " + name);
+        }
+    }
+
+    void run(OSDFPaths paths, ClusterCLI cli);
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/chaos/types/Chaos.java:45]: not sure that paths and cli should be part of an interface

File: src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+public class IOChaos implements Chaos {
+    @Getter
+    private final String name;
+    @Getter
+    private final List<String> components;
+    private final Integer severity;
+    private final Integer timeout;
+
+    @EqualsAndHashCode.Exclude
+    Random r = new Random();
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java:33]: why not private final?

File: src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+            throw new OSDFException("Timeout is incorrect or missing in " + name);
+        }
+        if (components == null || components.isEmpty()) {
+            throw new OSDFException("List of components is empty or missing in : " + name);
+        }
+    }
+
+    @Override
+    public String type() {
+        return "io";
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java:91]: Maybe create enum

File: src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
+package io.microconfig.osdf.deployers;
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/deployers/NetworkChaosDeployer.java:1]: Shouldn't be in deployers

File: src/main/java/io/microconfig/osdf/istio/Fault.java [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+import java.util.Map;
+
+import static io.microconfig.osdf.utils.YamlUtils.getInt;
+import static io.microconfig.osdf.utils.YamlUtils.getString;
+import static java.util.Map.of;
+
+
+@AllArgsConstructor
+public class Fault {
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/istio/Fault.java:14]: this class can be split

File: src/main/java/io/microconfig/osdf/utils/IstioUtils.java [hunk 20]
```diff
@@ -0,0 +1,20 @@
+package io.microconfig.osdf.utils;
+
+import io.microconfig.osdf.exceptions.OSDFException;
+import io.microconfig.osdf.service.ClusterService;
+import io.microconfig.osdf.service.istio.IstioService;
+
+public class IstioUtils {
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/utils/IstioUtils.java:7]: Bad utility class. Maybe move to IstioService interface

File: src/main/java/io/microconfig/osdf/paths/OSDFPaths.java [hunk 21]
```diff
@@ -45,13 +45,17 @@ public Path componentsPath() {
...
-        return of( configsPath() + "/components/system/versions/project-version.proc");
+        return of(configsPath() + "/components/system/versions/project-version.proc");
     }
 
     public Path configVersionPath() {
         return of(configsPath() + "/components/system/versions/config-version.proc");
     }
 
+    public Path chaosPlanPath() {
+        return of(componentsPath() + "/chaos/application.yaml");
```
Comment (SniXosha) [inline src/main/java/io/microconfig/osdf/paths/OSDFPaths.java:56]: component shouldn't be fixed

Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/paths/OSDFPaths.java]: not used, deleted

File: src/main/java/io/microconfig/osdf/chaos/validators/PodAndIOChaosIntersectionValidator.java [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+
+import static io.microconfig.osdf.chaos.types.ChaosType.IO;
+import static io.microconfig.osdf.chaos.types.ChaosType.POD;
+import static java.util.Collections.disjoint;
+import static java.util.stream.Collectors.toUnmodifiableSet;
+
+public class PodAndIOChaosIntersectionValidator {
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/validators/PodAndIOChaosIntersectionValidator.java:17]: fixed

File: src/main/java/io/microconfig/osdf/chaos/validators/BasicValidator.java [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static io.microconfig.osdf.chaos.types.ChaosType.NETWORK;
+import static io.microconfig.osdf.chaos.types.ChaosType.POD;
+import static java.util.stream.Collectors.toUnmodifiableSet;
+
+public class BasicValidator {
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/validators/BasicValidator.java:18]: fixed

File: src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+                break;
+            case PERCENT:
+                percentRun();
+                break;
+            default:
+                break;
+        }
+    }
+
+    private void percentRun() {
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java]: fixed

File: src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+
+    @SuppressWarnings("unchecked")
+    public static List<Chaos> parameterizedIOChaos(OSDFPaths paths, ClusterCLI cli, Entry<String, Object> entry, DurationParams durationParams) {
+        String name = entry.getKey();
+        Map<String, Object> yaml = (Map<String, Object>) entry.getValue();
+        List<String> components = (List<String>) (Object) getList(yaml, "components");
+        List<Integer> severities = intParamToList(getObjectOrNull(yaml, "params", "severity"), durationParams.getStagesNum());
+        ChaosMode mode = valueOf(getString(yaml, "mode").toUpperCase());
+        List<Chaos> chaosList = new ArrayList<>();
+        for (int i = 0; i < durationParams.getStagesNum(); i++) {
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/types/IOChaos.java]: fixed

File: src/main/java/io/microconfig/osdf/chaos/types/Chaos.java [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+    static List<Chaos> getAllChaosImpls(OSDFPaths paths, ClusterCLI cli) {
+        return List.of(
+                emptyIoChaos(paths, cli),
+                emptyNetworkChaos(paths, cli),
+                emptyPodChaos(paths, cli)
+        );
+    }
+
+    @SuppressWarnings("unchecked")
+    static List<Chaos> parameterizedChaosList(OSDFPaths paths, ClusterCLI cli, Entry<String, Object> entry, DurationParams durationParams) {
```
[the remark this answers is not in the document]
Comment (nvprotas) [reply src/main/java/io/microconfig/osdf/chaos/types/Chaos.java]: removed all except the first

**Pull request merged**